### PR TITLE
feat(eslint-plugin): [consistent-type-exports] check `export *` exports to see if all exported members are types

### LINKED ISSUE #9986: Enhancement: [consistent-type-exports] Check * exports to see if all exported values are types
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/consistent-type-exports/

### Description

Currently, consistent-type-exports only checked named exports and re-exports. However, sometimes we might write a file that only contains types & interfaces. It would be good to enforce using `export type *` instead of `export *` if all the exports are types.

### Fail

```TypeScript
// Component.props.ts
export interface Props {}

// Component.ts
export function Component(props: Props) {}

// index.ts
export * from './Component.props'
export * from './Component'
```

### Pass

```TypeScript
// Component.props.ts
export interface Props {}

// Component.ts
export function Component(props: Props) {}

// index.ts
export type * from './Component.props' // changed
export * from './Component'
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by auvred**
## PR Checklist

- [x] Addresses an existing open issue: fixes #9986
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

### CODE REVIEW

File: packages/eslint-plugin/src/rules/consistent-type-exports.ts [hunk 1]
```diff
@@ -77,34 +78,116 @@ export default createRule<Options, MessageIds>({
...
+        // TS checker has an internal function getExportsOfModuleWorker that
+        // recursively visits all module exports, including "export *". It then
+        // puts type-only-star-exported symbols into the typeOnlyExportStarMap
+        // property of sourceFile's SymbolLinks. Since symbol links aren't
+        // exposed outside the checker, we cannot access it directly.
+        //
+        // Therefore, to filter out value properties, we use the following hack:
+        // checker.getPropertiesOfType returns all exports that were originally
+        // values, but checker.getPropertyOfType returns undefined for
+        // properties that are mentioned in the typeOnlyExportStarMap.
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/consistent-type-exports.ts:156]: [Praise] Lovely. Nice investigation and very helpful comment! 🙂

**Review by JoshuaKGoldberg — APPROVED** [review]
🔥 !

**Comment by monodop** [comment]
Is Typescript 4 still supported on this project? This change introduces a dependency on typescript 5.0, which breaks Typescript 4 scripts that are running eslint via the Node.JS API.

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#support-for-export-type-

**Comment by JoshuaKGoldberg** [comment]
@monodop hmmmm. TS 4 should still be supported. Build tooling should be transpiling into `.d.ts` files supported by TS 4.8 per https://typescript-eslint.io/users/dependency-versions#typescript. If not, that's a bug we'd need to fix, so we'd request you please file an issue with a repro we can take a look at.

**Comment by monodop** [comment]
@JoshuaKGoldberg I think the easiest way to see the issue is to just look at the dist output. e.g. https://www.npmjs.com/package/@typescript-eslint/scope-manager?activeTab=code

image

**Comment by JoshuaKGoldberg** [comment]
#10065

**Pull request merged**